platform-core-module/src/main/java/com/softicar/platform/core/module/file/smb/jcifsng/JcifsNgSmbDirectory.java

```diff
@@ -1,74 +1,100 @@
 package com.softicar.platform.core.module.file.smb.jcifsng;
 
-import com.softicar.platform.common.core.exceptions.SofticarException;
 import com.softicar.platform.common.core.exceptions.SofticarIOException;
+import com.softicar.platform.common.core.interfaces.Predicates;
+import com.softicar.platform.common.string.Trim;
 import com.softicar.platform.core.module.file.smb.ISmbDirectory;
+import com.softicar.platform.core.module.file.smb.ISmbEntry;
 import com.softicar.platform.core.module.file.smb.ISmbFile;
-import java.io.InputStream;
+import com.softicar.platform.core.module.file.smb.SmbNoDirectoryException;
+import java.net.MalformedURLException;
 import java.util.ArrayList;
 import java.util.Arrays;
-import java.util.Collection;
 import java.util.List;
+import java.util.function.Function;
+import java.util.function.Predicate;
 import java.util.stream.Collectors;
 import jcifs.CIFSContext;
 import jcifs.SmbResource;
 import jcifs.smb.SmbException;
 import jcifs.smb.SmbFile;
 
-class JcifsNgSmbDirectory extends JcifsNgSmbFile implements ISmbDirectory {
+class JcifsNgSmbDirectory extends JcifsNgSmbEntry implements ISmbDirectory {
 
-	public JcifsNgSmbDirectory(String url, CIFSContext context) {
+	public JcifsNgSmbDirectory(SmbResource parent, String name) {
 
-		super(appendSlashIfMissing(url), context);
+		super(parent, appendSlashIfMissing(name));
+		assertDirectory();
 	}
 
-	public JcifsNgSmbDirectory(SmbResource parent, String name) {
+	public JcifsNgSmbDirectory(String url, CIFSContext context) {
 
-		super(parent, appendSlashIfMissing(name));
+		super(appendSlashIfMissing(url), context);
+		assertDirectory();
 	}
 
 	@Override
-	public List<ISmbFile> listFiles() {
+	public void makeDirectories() {
 
 		try {
-			return Arrays//
-				.asList(file.listFiles())
-				.stream()
-				.map(this::wrapSmbFile)
-				.collect(Collectors.toList());
+			if (!entry.exists()) {
+				entry.mkdirs();
+			}
 		} catch (SmbException exception) {
 			throw new SofticarIOException(exception);
 		}
 	}
 
 	@Override
-	public Collection<String> listFilesRecursively() {
+	public List<ISmbFile> listFiles() {
 
-		return listFiles("", file, new ArrayList<>());
+		return list(this::isFile, this::wrapFile);
 	}
 
 	@Override
-	public void mkdirs() {
+	public List<ISmbFile> listFilesRecursively() {
 
-		try {
-			if (!file.exists()) {
-				file.mkdirs();
-			}
-		} catch (SmbException exception) {
-			throw new SofticarIOException(exception);
+		List<ISmbFile> files = new ArrayList<>();
+		files.addAll(listFiles());
+		for (ISmbDirectory subDirectory: listSubDirectories()) {
+			files.addAll(subDirectory.listFilesRecursively());
 		}
+		return files;
+	}
+
+	@Override
+	public List<ISmbDirectory> listSubDirectories() {
```

Comment:
method should name `listSubdirectories()`

Reply:
I personally veto this, `listSubDirectories` has way better readability

Reply:
see discussion above - closing this one due to symmetry.